data/core/process.lua

```diff
@@ -0,0 +1,87 @@
+local config = require "core.config"
+
+process.stream = {}
+process.stream.__index = proc.stream
+
+function process.stream.new(proc, fd)
+  return setmetatable({ fd = fd, process = proc, buf = {}, len = 0 }, proc.stream)
+end
+
+-- read function can take either "*all", "*line", or a byte count.
+function process.stream:read(bytes, options)
+  options = options or {}
+  local start = system.get_time()
+  local target = bytes
+  if bytes == "*line" then
```

Comment:
might want to align something closer to Lua 5.4 not fully, just the initials instead of full lines):

```
"n": reads a numeral and returns it as a float or an integer, following the lexical conventions of Lua. (The numeral may have leading whitespaces and a sign.) This format always reads the longest input sequence that is a valid prefix for a numeral; if that prefix does not form a valid numeral (e.g., an empty string, "0x", or "3.4e-") or it is too long (more than 200 characters), it is discarded and the format returns fail.
"a": reads the whole file, starting at the current position. On end of file, it returns the empty string; this format never fails.
"l": reads the next line skipping the end of line, returning fail on end of file. This is the default format.
"L": reads the next line keeping the end-of-line character (if present), returning fail on end of file.
number: reads a string with up to this number of bytes, returning fail on end of file. If number is zero, it reads nothing and returns an empty string, or fail on end of file.
```

IIRC 5.4 also supports prefixing `*` in front, but its getting deprecated.

Reply:
OK, I'll put all that in. I'm going to leave off `"n"` though, as it's complicated to implement, and I don't think particularly useful in this interface. I'll just have it error if you try.